README.md
The 'FunChisq' R package
===============================

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/FunChisq)](https://cran.r-project.org/package=FunChisq)
[![CRAN_latest_release_date](https://www.r-pkg.org/badges/last-release/FunChisq)](https://cran.r-project.org/package=FunChisq)
[![metacran downloads](https://cranlogs.r-pkg.org/badges/FunChisq)](https://cran.r-project.org/package=FunChisq)
[![metacran downloads](https://cranlogs.r-pkg.org/badges/grand-total/FunChisq)](https://cran.r-project.org/package=FunChisq)



### Overview

The package provides statistical hypothesis testing methods for inferring model-free functional dependency. Functional test statistics are asymmetric and functionally optimal, unique from other related statistics. The test significance is based on either asymptotic chi-squared or exact distributions.

The tests include asymptotic *functional chi-squared tests* (Zhang & Song, 2013) [<arXiv:1311.2707>](https://arxiv.org/pdf/1311.2707v3.pdf) and an *exact functional test* (Zhong & Song, 2019) [<10.1109/TCBB.2018.2809743>](https://doi.org/10.1109/TCBB.2018.2809743) (Nguyen et al., 2020) [<10.24963/ijcai.2020/372>](https://doi.org/10.24963/ijcai.2020/372). The *normalized* functional chi-squared test was used by Best Performer NMSUSongLab in HPN-DREAM (DREAM8) Breast Cancer Network Inference Challenges (Hill et al., 2016) [<10.1038/nmeth.3773>](https://doi.org/10.1038/nmeth.3773). 

To measure the effect size, one can use the asymmetric *function index* (Zhong & Song, 2019) [<10.1186/s12920-019-0565-9>](https://doi.org/10.1186/s12920-019-0565-9) (Kumar et al., 2018) [<10.1109/BIBM.2018.8621502>](https://doi.org/10.1109/BIBM.2018.8621502). Its value is minimized to 0 by perfectly independent patterns and maximized to 1 by perfect non-constant functions.

A simulator (Sharma et al., 2017) [<10.32614/RJ-2017-053>](https://doi.org/10.32614/RJ-2017-053) can generate functional, non-functional, and independent patterns as contingency tables. The simulator provides options to control row and column marginal distributions and the noise level.

### When to use the package

Tests in this package can be used to reveal evidence for causality based on the causality-by-functionality principle. They target model-free inference without assuming a parametric model. For continuous data, these tests offer an advantage over regression analysis when a parametric functional form cannot be assumed. Data can be first discretized, e.g., by R packages ['Ckmeans.1d.dp'](https://cran.r-project.org/package=Ckmeans.1d.dp) or ['GridOnClusters'](https://cran.r-project.org/package=GridOnClusters). For categorical data, they provide a novel means to assess directional dependency not possible with symmetrical Pearson's chi-squared or Fisher's exact tests. They are a better alternative to conditional entropy in many aspects.

### To download and install the package

```{r}
install.packages("FunChisq")
```


NEWS
NEWS

Version 2.5.2

  2021-05-19
  1. Created 2.5.2 from 2.5.1.2

Version 2.5.1.2 (not publicly released)

  2021-05-17
  1. Updated the is_many.to.one() function in simulate_tables() to
  satisfy the non-monotonic property of functional patterns within
  the rows and columns having non-zero marginals.

  2. Added a new method option "adapted" to fun.chisq.test.

  3. Added a vignette 'adapted.fun.chisq.test.Rmd' for adapted
  functional chi-squared test.

  4. Mentioned "adapted functional chi-squared test" in DESCRIPTION.

  5. Imported R packages 'dqrng', 'DescTools' and 'infotheo' in
  DESCRIPTION.

  6. Added three testthat test cases for adapted functional
  chi-squared test in test_AdpFunChisq.R.

  7. Updated fun.chisq.test.Rd with details about the method option
  "adapted".

  8. Added imports from 'dqrng', 'infotheo' and 'DescTools' in NAMESPACE.

  2021-05-11
  1. Created 2.5.1.2 from 2.5.1.1

  2. Updated the not_constant() function in simulate_tables() to
  satisfy the non-constant property of functional patterns within
  the rows and columns having non-zero marginals.

Version 2.5.1.1 (not publicly released)

  2021-04-18
  1. Removed the 'LazyData: TRUE' option in the package DESCRIPTION
  as the package does not have a 'data' directory.

  2021-03-26
  1. Further improved numerical stability for calculating FunChisq
  statistic and function index (estimate) when alternative =
  "non-constant" (default) for fun.chisq.test().
  2. Added two more testthat cases for in test_FunChisq.R

  2021-03-24
  1. Fixed a bug that if the input has all zeros, the
  fun.chisq.test() function would stop with an error. A test case
  is added to the testthat examples.
  2. Fixed a bug that if an input table has small non-integer values,
  round-off errors can result in calculating function index
  (estimate), now the issue has been fixed. A test case is added
  to the testthat examples.

  2020-11-27
  1. Fixed the p.value entry to be unamed in the return object of
     fun.chisq.test(). It was incorrectly named "statistic" in some
     test options. The change was made inside file "R/FunChisq.R".

  2020-07-18
  1. Updated REFERENCES.bib and CITATION
  2. Updated README.md

  2020-06-14
  1. Created 2.5.1.1 from 2.5.1

Version 2.5.1

  2020-06-13
  1. Created 2.5.1 from 2.5.0.1

Version 2.5.0.1 (not publicly released)

  2020-06-13
  1. Updated the README file to include badges about the package.
  2. Replaced \dontrun by \donttest for examples in manuals.

  2020-05-31
  1. API change to simulate_tables(). For functional patterns, if
     both row and column marginal distributions are specified, each
     marginal is observed half of the time with the other marginal
     not strictly observed. In the past, only the row marginal
     distribution was observed.
  2. The simulate_tables() function now can generate functional
     patterns for a given column marginal distribution
  3. simulate_tables(): Default row and column marginals are changed
     to NULL at the parameter. Any NULL is replaced by uniform
     distribution if not specified by user.
  4. simulate_tables(): Introduced decide.marginal() to select the
     marginal type if both are provided or none of them are provided.
  5. simulate_tables(): Changed the name of functional.table()
     function to disc.functional.pattern().
  6. simulate_tables(): Functional 0/1 pattern table is no longer
     dependent on either row and column marginal.
  7. Added new testthat tests for simulate_tables() to check for
     marginal distribution of simulated tables.
  8. Added three more examples of functional tables in the
     simulate_tables() manual with user specified column marginals.
  9. Updated simulate_tables() manual.

  2020-04-24
  1. Created version 2.5.0.1 from 2.5.0.
  2. Updated manuals of most functions in the package.

Version 2.5.0

  2020-04-23
  1. API change: simulate_tables() now includes a margin parameter to
     specify how noise model will be applied. The default noise for
     functional patterns has now been changed to apply along both row
     and column. In previous versions, the noise was only applied
     along columns within each row such that row sums do not change.
  2. Test_Functional_table() has been modified to not check for
     equal row sums before and after adding noise, which is true only
     if the noise is applied along the column.
  3. Updated references, CITATION, and manuals to reflect a newly
     accepted manuscript (Nguyen et al. 2020).

  2020-03-15
  1. Fixed a NOTE in 2.4.9.2 Rdpack was not imported in NAMESPACE.
  2. Updated references in FunChisq-package.Rd and REFERENCES.bib.

  2020-03-14
  1. Created version 2.5.0 from 2.4.9.2.

Version 2.4.9.2

  2020-03-13
  1. Updated CITATION, DESCRIPTION, README.md
  2. Introduced the Rdpack to manage references

  2020-02-13
  1. Added ORCID for author Sajal Kumar.

  2019-11-10
  1. Updated the REAMDE.md file of the package.

  2019-11-04
  2. Fixed "page" entry to "pages" entry in the CITATION file.
  3. Deprecated 'cp.chisq.test'. The function has been moved to R
     package 'DiffXTables'.

  2019-09-28
  1. Created version 2.4.9.2.
  2. Fixed signed / unsigned mismatch in trimTable.cpp.

Version 2.4.9.1

  2019-09-20
  1. Created version 2.4.9.1.
  2. Fixed a bug in trimming all-zero rows or columns from the
     input table in EFT DQP and DP. Added a new test case x16 for
     this bug.
  3. Eliminated the use of static array in the EFT DP and DQP
     code.

Version 2.4.9 (not publicly released)

  2019-09-08
  1. Created version 2.4.9.
  2. Revised README.md.
  3. Fixed a bug for EFT DQP. Included the table that caused the
  bug as test case table x15.

Version 2.4.8-1

  2019-09-08
  1. Revised 'DESCRITION'.
  2. Revised 'README.md'.

  2019-09-07
  1. Fixed a bug in the exact.dqp option of the exact functional
  test. Added a new test case that had given this bug in the test
  functions.

R/FunChisq.R
# FunChisq.R -- statistical tests for model-free functional dependency
#
# YZ, HZ, MS
# Modified: Feb 8, 2015; June 25, 2015
# Jan 22, 2016 MS:
#    1. introduced the type argument to indicate whether alternative hypothesis is functional
#       or non-constant function
#    2. introduced the log.p argument to obtain the log of p-value
# Jan 23, 2016 MS:
#    Added a return of an estimate of functional index between 0 and 1. It is asymetrical and
#      different from Cramer's V.
# Jan 30, 2016 MS:
#    Added a new argument "index.kind" to specify the function index kind: "unconditional"
#      or "conditional" given marginal of Y
# Feb 5, 2016 MS:
#    Renamed the "type" argument to "alternative"
# Feb 7, 2016 MS:
#   Handled the special case when the degrees of freedom for the normalized FunChisq are zero
# May 3, 2016 MS:
#   Listed all arguments
# May 29, 2017 HZ:
#   Added "simulate.p.value" method to calculate p-value with simulated distribution
# Feb 08, 2018 HZ:
#   Added exact functional test option "exact.mode.bound"
# Oct 24, 2018 MS:
#   Default to (conditional) function index. The previous default was unconditional function
#   index.

fun.chisq.test <- function (
  x,
  method=c("fchisq", "nfchisq", "exact", "default",
           "normalized", "simulate.p.value"),
  alternative=c("non-constant", "all"), log.p=FALSE,
  index.kind=c("conditional", "unconditional"
#               , "fix.row.sums", "fix.column.sums",
#               "fix.marginal.sums"
               ),
  simulate.nruns = 2000,
  exact.mode.bound = TRUE
)
{
  if(!is.matrix(x) && !is.data.frame(x)) stop("input x must be matrix or data frame\n")

  method <- match.arg(method)
  if(method == "default") {
    warning(paste0("method=\"", method, "\" is deprecated. Use \"fchisq\" instead."))
    method <- "fchisq"
  } else if(method == "normalized") {
    warning(paste0("method=\"", method, "\" is deprecated. Use \"nfchisq\" instead."))
    method <- "nfchisq"
  }

  alternative <- match.arg(alternative)
  index.kind <- match.arg(index.kind)

  row.sums <- rowSums(x)
  col.sums <- colSums(x)

  n <- sum(col.sums)

  if(0) {
    row.chisq.sum <- sum(apply(x, 1,
                               function(v){
                                 row.sum <- sum(v)
                                 expected <- row.sum / length(v)
                                 if(row.sum>0) sum( (v - expected)^2 / expected)
                                 else 0
                               }))
  } else { # MS June 13, 2017
    row.chisq.sum <- sum(sapply(seq(nrow(x)),
                                function(i){
                                  if(row.sums[i]>0) sum( x[i, ]^2 ) / row.sums[i]
                                  else 0
                                })) * ncol(x) - n
  }

  if(alternative == "non-constant") { # non-constant functional chi-square
    # col.expected <- n / ncol(x)
    # col.chisq <- sum((col.sums - col.expected)^2 / col.expected)

    col.chisq <- sum(col.sums^2) / n * ncol(x) - n

    fun.chisq <- row.chisq.sum - col.chisq
    df <- nrow(x) * (ncol(x) - 1) - (ncol(x) - 1)

    # Version 1: Current. Added June 11, 2017. Bound is maximum reachable
    if(index.kind == "unconditional") { # unconditional functional index
      max.fun.chisq <- n * ncol(x) * (1 - 1.0 / min(dim(x)))
      estimate.label <- "unconditional function index xi.f"
    } else if(index.kind == "conditional" || index.kind == "fix.column.sums") {
      # functional index given the column marginal

      # max.fun.chisq <- sum(sort(col.sums, decreasing=TRUE)[seq(min(dim(x)))]) *
      #  ncol(x) - n - col.chisq

      max.fun.chisq <- n * ncol(x) - n - col.chisq

      estimate.label <- "function index xi.f" # "function index xi.f conditioned on column sums"

    } else if(index.kind == "fix.row.sums") { # Y-conditional functional index

      max.fun.chisq <- sum(sort(row.sums, decreasing=TRUE)[seq(min(dim(x)))]) *
        ncol(x) - n - col.chisq
      estimate.label <- "function index xi.f conditioned on row sums (inaccurate)"

    } else if(index.kind == "fix.marginal.sums") { # XY-conditional functional index

      max.fc.given.col.sums <- sum(sort(col.sums, decreasing=TRUE)[seq(min(dim(x)))]) *
        ncol(x) - n - col.chisq

      max.fc.given.row.sums <- sum(sort(row.sums, decreasing=TRUE)[seq(min(dim(x)))]) *
        ncol(x) - n - col.chisq

      max.fun.chisq <- min(c(max.fc.given.col.sums, max.fc.given.row.sums))

      estimate.label <- "function index xi.f conditioned on row and column sums (inaccurate)"
    }

    # Version 0: bound is not reachable when nrow(x) < ncol(x)
    else if(index.kind == "conditional-version-0") { # functional index given the column marginal
      max.fun.chisq <- n * (ncol(x) - 1) - col.chisq
      estimate.label <- "conditional function index xi.f"
    } else if(index.kind == "unconditional-version-0") { # unconditional functional index
      max.fun.chisq <- n * (ncol(x) - 1)
      estimate.label <- "function index xi.f"
    } else {
      stop("unrecognized function index kind ", index.kind)
    }

  } else if(alternative == "all") { # functional chi-square

    fun.chisq <- row.chisq.sum
    df <- nrow(x) * (ncol(x) - 1)

    if(index.kind == "unconditional") {
      max.fun.chisq <- n * (ncol(x) - 1)
      estimate.label <- "function index xi.f"

    } else if(index.kind == "conditional" || index.kind == "fix.column.sums") {
      # MS. Added June 11, 2017
      # max.fun.chisq <- sum(sort(col.sums, decreasing=TRUE)[seq(min(dim(x)))]) *
      #  ncol(x) - n

      max.fun.chisq <- n * ncol(x) - n

      estimate.label <- "function index xi.f conditioned on column sums"

    } else if(index.kind == "fix.row.sums") { # Y-conditional functional index

      max.fun.chisq <- sum(sort(row.sums, decreasing=TRUE)[seq(min(dim(x)))]) *
        ncol(x) - n
      estimate.label <- "function index xi.f conditioned on row sums"

    } else if(index.kind == "fix.marginal.sums") { # XY-conditional functional index

      max.fc.given.col.sums <- sum(sort(col.sums, decreasing=TRUE)[seq(min(dim(x)))]) *
        ncol(x) - n

      max.fc.given.row.sums <- sum(sort(row.sums, decreasing=TRUE)[seq(min(dim(x)))]) *
        ncol(x) - n

      max.fun.chisq <- min(c(max.fc.given.col.sums, max.fc.given.row.sums))
      estimate.label <- "function index xi.f conditioned on row and column sums"
    }

  } else {
    stop("unknown alternative ", alternative)
  }

  estimate.label <- paste(alternative, estimate.label)

  if(max.fun.chisq > 0) {
    estimate <- sqrt(abs(fun.chisq) / max.fun.chisq)
  } else {
    estimate <- 0
  }

  names(estimate) <- estimate.label

  DNAME <- deparse(substitute(x))

  if(method=="fchisq") {
    method.text <- "Functional chi-square test"
    names(fun.chisq) <- "statistic"
    names(df) <- "parameter"
    p.value <- pchisq( fun.chisq, df = df, lower.tail=FALSE, log.p=log.p )
    return(structure(list( statistic=fun.chisq, parameter=df, p.value=p.value,
                           estimate = estimate, data.name=DNAME,
                           method = method.text),
                     class = "htest"))

  } else if(method=="nfchisq") {
    method.text <- "Normalized functional chi-square test"
    if(df > 0) {
      normalized <- as.numeric((fun.chisq-df)/sqrt(2*df))
    } else {
      normalized <- -Inf
    }
    p.value <- pnorm( normalized, lower.tail=FALSE, log.p=log.p )
    names(normalized) <- "statistic"
    names(df) <- "parameter"
    return(structure(list(statistic = normalized, parameter = df, p.value = p.value,
                          estimate = estimate, data.name = DNAME, method = method.text),
                     class = "htest"))
  } else if(method=="simulate.p.value"){
    method.text <- "Functional chi-square test with simulated p value"
    names(fun.chisq) <- "statistic"
    names(df) <- "parameter"
    #p.value <- pchisq( fun.chisq, df = df, lower.tail=FALSE, log.p=log.p )
    p.value <- simulate.p.value(x, simulate.nruns)

    return(structure(list( statistic=fun.chisq, parameter=df, p.value=p.value,
                           estimate = estimate, data.name=DNAME,
                           method = method.text),
                     class = "htest"))
  } else if(method=="exact") {

    method.text <- "Exact functional test"
    if(sum(x%%1!=0)>=1) { # Check whether numbers in x are all integers
      stop("ERROR: Exact test requires integer contingency tables!", call. = TRUE)
    }
    ####

    ####
    #Hua added, Nov 13, 2014
    #Exact functional test
    if((sum(x) <= 200 || sum(x)/nrow(x)/ncol(x) <=5)
       && nrow(x)<=10 && ncol(x)<=10) {
      # p.value <- exact.functional.test(x)
      p.value <- ExactFunctionalTest(x, exact.mode.bound)
      if(log.p) p.value <- log(p.value)
      names(fun.chisq) <- "statistic"
      return(structure(list(statistic = fun.chisq, p.value = p.value, estimate = estimate,
                            data.name = DNAME, method = method.text),
                       class = "htest"))
    } else {
      warning("Asymptotic test is used in place of exact test to save time.\n")
      return(fun.chisq.test(x, method="fchisq", alternative=alternative, log.p=log.p,
                            index.kind=index.kind))
    }

    ####
  } else {
    stop("ERROR: unrecognized method argument", method)
  }
}
